feat(api): add buf filter to nvim_list_chans()

benarcher2691 opened the pull request:
## Summary

Add optional `{ buf = N }` parameter to `nvim_list_chans()` to filter channels by terminal buffer. This follows the pattern planned for all "get" APIs.

- Add `Dict(list_chans)` keyset with optional `buf` field
- Update `channel_all_info()` to accept filter parameter
- Update existing test cases to pass empty dict `{}` (required by new signature)

## Example

```lua
vim.api.nvim_create_autocmd('TermOpen', {
  callback = function(ev)
    local chans = vim.api.nvim_list_chans({ buf = ev.buf })
    if chans[1] and #chans[1].argv <= 1 then
      vim.cmd('startinsert')
    end
  end
})
```

## Test plan

- [x] Added test for `{ buf = N }` filter in vim_spec.lua
- [x] Updated existing tests to use `nvim_list_chans({})` 
- [x] All functional tests pass

Ref #36678

benarcher2691 commented:
I missed that while the api-contract allows adding an optional `opts` parameter, the current implementation actually requires passing `{}` at the call site - making this a breaking change.

I see #31903 tracks making `opts` truly optional, but it's not implemented yet.

How would you suggest proceeding here? Should I wait for #31903, or take a different approach?

seandewar commented:
Historically we'd introduce something like `nvim_list_chans2` and deprecate the old one, but with #31903 open it's probably better to wait for that to be addressed instead.

Open PR #33392 exists to relax `opts`, but seems stalled; feel free to take over that work if you're up to it.

In any case, I'll ping @justinmk for his opinion. :pray:

yochem commented:
See this command by Justin: https://github.com/neovim/neovim/pull/27195#pullrequestreview-1847670314 and the relevant docs in `:help dev-naming`:

> get: Gets things. Two variants (overloads): 1. `get<T>(id: int): T` returns one item. 2. `get<T>(filter: dict): T[]` returns a list.

> Do NOT use these deprecated verbs:
> - ...
> - list: Redundant with "get"

Thus creating a `nvim_chan_get` function and deprecating `nvim_get_chan_info` and `nvim_list_chans` seems like the better approach. Justin proposed the signature `nvim_chan_get(id, opts)`, however, I would go with just `nvim_chan_get(opts)` and have `id` be one of the keys in `opts`, similar to `nvim_get_autocmds()`.

benarcher2691 committed:
feat(api): add nvim_chan_get() unified channel API

Add nvim_chan_get(opts) as a unified replacement for nvim_get_chan_info()
and nvim_list_chans(), following the pattern established by
nvim_get_autocmds().

Options:
- id: (integer) Filter by channel id
- buf: (integer) Filter by terminal buffer handle

Returns an array of channel info dictionaries.

Deprecate nvim_get_chan_info() and nvim_list_chans() in favor of the
new unified API.

Ref #36678
